crates/turborepo-lib/src/process/mod.rs

```diff
@@ -0,0 +1,193 @@
+//! `process`
+//!
+//! This module contains the code that is responsible for running the commands
+//! that are queued by run. It consists of a set of child processes that are
+//! spawned and managed by the manager. The manager is responsible for
+//! running these processes to completion, forwarding signals, and closing
+//! them when the manager is closed.
+//!
+//! As of now, the manager will execute futures in a random order, and
+//! must be either `wait`ed on or `stop`ped to drive state.
+
+mod child;
+mod shared_adaptor;
+
+use std::time::Duration;
+
+pub use child::Command;
+pub use shared_adaptor::SharedProcessManager;
+use tokio::task::JoinSet;
+use tracing::{debug, trace};
+
+/// A process manager that is responsible for spawning and managing child
+/// processes. When the manager is Open, new child processes can be spawned
+/// using `spawn`. When the manager is Closed, all currently-running children
+/// will be closed, and no new children can be spawned.
+#[derive(Debug)]
+pub struct ProcessManager<T> {
+    state: T,
+}
+
+#[derive(Debug)]
+pub struct Open(Vec<child::Child>);
+#[derive(Debug)]
+pub struct Closed;
+
+impl ProcessManager<Closed> {
+    pub fn new() -> Self {
+        ProcessManager { state: Closed }
+    }
+
+    pub fn start(self) -> ProcessManager<Open> {
+        ProcessManager {
+            state: Open(Default::default()),
+        }
+    }
+}
+
+impl ProcessManager<Open> {
+    /// Spawn a new child process to run the given command.
+    ///
+    /// The handle of the child can be either waited or stopped by the caller,
+    /// as well as the entire process manager.
+    pub fn spawn(&mut self, command: child::Command, stop_timeout: Duration) -> child::Child {
+        let child = child::Child::spawn(command, child::ShutdownStyle::Graceful(stop_timeout));
+        self.state.0.push(child.clone());
+        child
+    }
+
+    /// Stop the process manager, closing all child processes. On posix
+    /// systems this will send a SIGINT, and on windows it will just kill
+    /// the process immediately.
+    pub async fn stop(self) -> ProcessManager<Closed> {
+        let mut set = JoinSet::new();
+
+        for mut child in self.state.0.into_iter() {
+            set.spawn(async move { child.stop().await });
+        }
+
+        debug!("waiting for {} processes to exit", set.len());
+
+        while let Some(out) = set.join_next().await {
```

Comment:
We should probably have a timeout on this, the Go code has a `killTimeout`

Reply:
There is a timeout per-child which should provide an upper bound, although there is perhaps a situation where kill hangs _also_. I may set the timeout here to be whatever the kill timeout is plus one second.